packages/core/src/defer/triggering.ts

```diff
@@ -397,14 +398,20 @@ export async function triggerHydrationFromBlockName(
   for (const dehydratedBlockId of hydrationQueue) {
     await triggerResourceLoadingForHydration(dehydratedBlockId, dehydratedBlockRegistry);
     await nextRender(injector);
-    // TODO(incremental-hydration): assert (in dev mode) that a defer block is present in the dehydrated registry
-    // at this point. If not - it means that the block has not been hydrated, for example due to different
-    // `@if` conditions on the client and the server. If we detect this case, we should also do the cleanup
-    // of all child block (promises, registry state, etc).
-    // TODO(incremental-hydration): call `rejectFn` when lDetails[DEFER_BLOCK_STATE] is `DeferBlockState.Error`.
-    blocksBeingHydrated.get(dehydratedBlockId)!.resolve();
-
-    // TODO(incremental-hydration): consider adding a wait for stability here
+    if (typeof ngDevMode === 'undefined' || ngDevMode) {
+      assertDeferBlockInRegistry(dehydratedBlockRegistry, dehydratedBlockId);
+      // TODO(incremental-hydration): the block has not been hydrated, for example due to different
+      // `@if` conditions on the client and the server. We should also do the cleanup of all child
+      // block (promises, registry state, etc).
+    }
+    const deferBlock = dehydratedBlockRegistry.get(dehydratedBlockId)!;
+    const lDetails = getLDeferBlockDetails(deferBlock.lView, deferBlock.tNode);
+    const hydratingBlockPromise = blocksBeingHydrated.get(dehydratedBlockId)!;
+
+    if (lDetails[DEFER_BLOCK_STATE] === DeferBlockState.Error) {
+      hydratingBlockPromise.reject();
+    }
+    hydratingBlockPromise.resolve();
```

Comment:
In a context of a previous comment (https://github.com/angular/angular/pull/59032/files#r1868731810), we can update the code like this:
```suggestion
    if (deferBlock) {
      const lDetails = getLDeferBlockDetails(deferBlock.lView, deferBlock.tNode);
      const hydratingBlockPromise = blocksBeingHydrated.get(dehydratedBlockId)!;

      const state = lDetails[DEFER_BLOCK_STATE];
      if (state === DeferBlockState.Error) {
        hydratingBlockPromise.reject();
      } else if (state === DeferBlockState.Complete) {
        hydratingBlockPromise.resolve();
      }
      // TODO(incremental-hydration): handle a case when the `state` is different from the ones we
      // handle above. Consider adding an assert (if that's an invalid state).
    } else {
      // TODO(incremental-hydration): handle a case when a defer block is not present in the registry
      // after hydration of parent block(s), for example due to different `@if` conditions on the client
      // and the server. We should also do the cleanup of all child block (promises, registry state, etc).
    }
```

Reply:
This has been moved to a separate branch / PR.